### PR TITLE
ENH: Add Similarity3DTransform in itkLandmarkBasedTransformInitializer

### PR DESCRIPTION
**Pull request by PranjalSahu**
For adding iso-tropic scaling in the transformation matrix along with Rotation and Translation when using landmarks for transform initialization.
Moved some common code from VersorRigid3DTransform and Similarity3DTransform and created **ComputeCentroid** and **CreateMatrix**.
Introduced a PointType3D to handle the compilation issue when wrapping.

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [X] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

### CODE REVIEW

**Review by thewtex — APPROVED** [review]
:100:

**Review by thewtex — CHANGES_REQUESTED** [review]
2D build error:

  https://open.cdash.org/viewBuildError.php?buildid=8072993

**Comment by PranjalSahu** [comment]
Yes, there is some issue with the Python wrapping caused due to code refactoring.
I am changing the .wrap file to account for it.

**Comment by PranjalSahu** [comment]
Introduced a PointType3D to handle the compilation issue when wrapping.

**Review by dzenanz — APPROVED** [review]
The changes looks fine. Does it compile? 😄

**Comment by PranjalSahu** [comment]
Yes I tested it locally with ctest. Doing one more test with a bigger dataset to check if it gives the correct output. 
With the smaller test sample, it gave the correct scaling output.

**Comment by dzenanz** [comment]
As Matt is on vacation this week, let's not wait for his review. If @blowekamp does not review until 3-4PM, I say we merge.

**Comment by PranjalSahu** [comment]
@dzenanz yes. I discovered one bug in mesh writing in binary mode in rc4.post2 that delayed the testing of this PR.
I will confirm here once that is done.

**Review by blowekamp — APPROVED** [review]
Looks good.

The formatting of the comments could be improved.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 1]
```diff
@@ -372,54 +379,172 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
+                                                                                 vnl_matrix<ParametersValueType>>;
+    SymmetricEigenAnalysisType symmetricEigenSystem;
+
+    symmetricEigenSystem.ComputeEigenValuesAndVectors(N, eigenValues, eigenVectors);
+
+    itkDebugMacro(<< "EigenVectors " << eigenVectors);
+    itkDebugMacro(<< "EigenValues " << eigenValues);
+
+    // By default eigen values are sorted in ascending order.  therefore the
+    // maximum
```
Comment (dzenanz) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:463]: Reflow this comment paragraph.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 2]
```diff
@@ -372,54 +379,172 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
+  {
+    itkExceptionMacro("Transform is Similiarity3DTransform and Fixed image dimension is not 3");
+  }
+  if (MovingImageType::ImageDimension != 3)
+  {
+    itkExceptionMacro("Transform is Similiarity3DTransform and Moving image dimension is not 3");
+  }
+
+  // --- compute the necessary transform to match the two sets of landmarks
+  // ---
```
Comment (dzenanz) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:510]: We don't need this much whitespace in this comment block.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 3]
```diff
@@ -372,54 +379,172 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
+
+  PointType3D movingCentroid = ComputeCentroid(m_MovingLandmarks);
 
   itkDebugMacro(<< "fixed centroid  = " << fixedCentroid);
   itkDebugMacro(<< "moving centroid  = " << movingCentroid);
 
   using VersorType = typename VersorRigid3DTransformType::VersorType;
 
   VersorType versor;
+  double     sa = 0.0, sb = 0.0;
```
Comment (dzenanz) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:538]: I think that ITK style suggest one variable declaration per line.

**Comment by PranjalSahu** [comment]
I have tested it using Python and there it is calling LandmarkBasedTransformInitializer with the VersorRigid3DTransform when using Similarity3DTransform. This is because Similarity3DTransform inherits from VersorRigid3DTransform. I will have to check how to handle this for Python wrappings.

**Comment by PranjalSahu** [comment]
Changing the ordering of dynamic_cast and made it work in Python.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.h [hunk 4]
```diff
@@ -192,6 +196,9 @@ class ITK_TEMPLATE_EXPORT LandmarkBasedTransformInitializer : public Object
   /** Initializer for VersorRigid3D */
   void
   InternalInitializeTransform(VersorRigid3DTransformType *);
+  /** Initializer for Similarity3DTransform */
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.h:199]: Although there are many other docstrings that do not end with a period (p. 373, 374), at least for the ones that are added/reworked, it'd be nice to add it following the ITK SW Guide.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 5]
```diff
@@ -355,71 +361,180 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
-  // quaternions,"
-  //    Journal of the Optical Society of America A, 4:629-642
-  //
-  //
-  //    Original python implementation by David G. Gobbi
-  //    Readpted from the code in VTK: Hybrid/vtkLandmarkTransform
-  //
-  //----------------------------------------------------------------------------
+  /*----------------------------------------------------------------------------
+     Compute the necessary transform to match the two sets of landmarks
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:365]: For comment blocks, the ITK SW guide recommends using // (p.374), so not sure why this was changed.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 6]
```diff
@@ -355,71 +361,180 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
-  using PointType = typename VersorRigid3DTransformType::OutputPointType;
 
   // Compute the centroids
-  PointType fixedCentroid;
-  fixedCentroid.Fill(0.0);
-  PointsContainerConstIterator fixedItr = m_FixedLandmarks.begin();
-  while (fixedItr != m_FixedLandmarks.end())
+  PointType3D fixedCentroid = ComputeCentroid(m_FixedLandmarks);
+
+  PointType3D movingCentroid = ComputeCentroid(m_MovingLandmarks);
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:381]: The ITK SW Guide recommends using `this->` to access member methods (p. 333) (applies to any code in the PR; have not checked all calls).

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 7]
```diff
@@ -355,71 +361,180 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
+    maximum eigen value is the one in the fourth place = index 3. We need the
+    eigen vector associated with the maximum eigenvalue, so we take the
+    eigenvector from the last row, index=3 */
+
+    versor.Set(eigenVectors[3][1], eigenVectors[3][2], eigenVectors[3][3], eigenVectors[3][0]);
+    itkDebugMacro(<< "Resulting versor" << versor);
+  }
+  else
+  {
+    // Remember..
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:467]: Not sure this comment is necessary, or whether the else block as a whole is necessary, as it is empty. There is already a mention to the need of having at least 3 landmarks at the beginning of the `if` block.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 8]
```diff
@@ -355,71 +361,180 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
+     Compute the necessary transform to match the two sets of landmarks
+
+     The solution is based on
+     Berthold K. P. Horn (1987),
+     "Closed-form solution of absolute orientation using unit quaternions,"
+     Journal of the Optical Society of America A, 4:629-642
+
+     Original python implementation by David G. Gobbi
+     Readapted from the code in VTK: Hybrid/vtkLandmarkTransform
+  ----------------------------------------------------------------------------*/
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:511]: Same comment as above.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 9]
```diff
@@ -438,6 +553,12 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
         movingCentered[i] = (*movingItr)[i] - movingCentroid[i];
       }
 
+      // accumulate scale factors (if desired)
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:556]: Start with capitals (ITK SW Guide).

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 10]
```diff
@@ -491,13 +599,10 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
-    // maximum
-    // eigen value is the one  in the fourth place = index 3. We need the
-    // eigen
-    // vector associated with the maximum eigenvalue, so we take the
-    // eigenvector
-    // from the last row, index=3.
+    /* By default eigen values are sorted in ascending order therefore the
+    maximum eigen value is the one  in the fourth place = index 3. We need the
+    eigen vector associated with the maximum eigenvalue, so we take the
+    eigenvector from the last row, index=3. */
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:605]: Same comment as above about comment block style.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 11]
```diff
@@ -666,6 +779,51 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Initia
...
+  centroid[2] /= inputLandmarks.size();
+
+  return centroid;
+}
+template <typename TTransform, typename TFixedImage, typename TMovingImage>
+void LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::CreateMatrix(
+  itk::Matrix<ParametersValueType, 4, 4> &                               N,
+  const itk::Matrix<ParametersValueType, ImageDimension, ImageDimension> M)
+{
+  // on-diagonal elements
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:810]: Start with capitals.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 12]
```diff
@@ -666,6 +779,51 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Initia
...
+  N[0][0] = M[0][0] + M[1][1] + M[2][2];
+  N[1][1] = M[0][0] - M[1][1] - M[2][2];
+  N[2][2] = -M[0][0] + M[1][1] - M[2][2];
+  N[3][3] = -M[0][0] - M[1][1] + M[2][2];
+  // off-diagonal elements
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:815]: Same as above.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 13]
```diff
@@ -355,71 +361,180 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
+      {
+        fixedCentered[i] = (*fixedItr)[i] - fixedCentroid[i];
+        movingCentered[i] = (*movingItr)[i] - movingCentroid[i];
+      }
+
+      for (unsigned int i = 0; i < ImageDimension; ++i)
+      {
+        for (unsigned int j = 0; j < ImageDimension; ++j)
+        {
+          // mmm this indices i,j may have to be reverted...
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:420]: Does this "mmm" mean more investigation is required? Can it be tested/checked through a test?

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx [hunk 14]
```diff
@@ -355,71 +361,180 @@ LandmarkBasedTransformInitializer<TTransform, TFixedImage, TMovingImage>::Intern
...
+
+      itkDebugStatement(++ii);
+      itkDebugMacro(<< "f_" << ii << " = " << fixedCentered);
+      itkDebugMacro(<< "m_" << ii << " = " << movingCentered);
+
+      ++movingItr;
+      ++fixedItr;
+    }
+
+    // -- build the 4x4 matrix N --
```
Comment (jhlegarreta) [inline Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:433]: Stick to the ITK SW Guide comment style guidelines.

**Review by jhlegarreta — COMMENTED** [review]
Some in-line comments, mainly to increase coding style consistency, and one that seems like a suspicious ToDo-style comment.

File: Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx (hunk 13 above)
Comment (PranjalSahu) [reply Modules/Registration/Common/include/itkLandmarkBasedTransformInitializer.hxx:420]: I have tested it with a bigger mesh also and it works. I will update the comment.

**Review by thewtex — APPROVED** [review]
:100:

**Pull request merged**